Про существующие инструменты хорошо написано вот тут  - http://web-corpora.net/wsgi/mystemplus.wsgi/mystemplus/instructions/

Не все из списка просто установить даже в колабе. Воспользуемся TreeTagger. Он хотя бы работает.

Скачиваем нужные файлы по инструкции.

In [2]:
!wget http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tree-tagger-linux-3.2.2.tar.gz

--2019-06-16 11:42:08--  http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tree-tagger-linux-3.2.2.tar.gz
Resolving www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)... 129.187.148.72, 2001:4ca0:4f01::5
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tree-tagger-linux-3.2.2.tar.gz [following]
--2019-06-16 11:42:09--  https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tree-tagger-linux-3.2.2.tar.gz
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1821366 (1.7M) [application/x-gzip]
Saving to: ‘tree-tagger-linux-3.2.2.tar.gz’

tree-tagger-linux-3 100%[===================>]   1.74M   995KB/s    in 1.8s    

2019-06-16 11:42:12 (995 KB/s) - ‘tree-tagger-linux-3.2.2.tar.gz’ saved [1821366/1821366]



In [3]:
!wget http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tagger-scripts.tar.gz

--2019-06-16 11:42:13--  http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tagger-scripts.tar.gz
Resolving www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)... 129.187.148.72, 2001:4ca0:4f01::5
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tagger-scripts.tar.gz [following]
--2019-06-16 11:42:14--  https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/tagger-scripts.tar.gz
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 101810 (99K) [application/x-gzip]
Saving to: ‘tagger-scripts.tar.gz’

tagger-scripts.tar. 100%[===================>]  99.42K   165KB/s    in 0.6s    

2019-06-16 11:42:16 (165 KB/s) - ‘tagger-scripts.tar.gz’ saved [101810/101810]



In [4]:
!wget http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/install-tagger.sh

--2019-06-16 11:42:19--  http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/install-tagger.sh
Resolving www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)... 129.187.148.72, 2001:4ca0:4f01::5
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/install-tagger.sh [following]
--2019-06-16 11:42:19--  https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/install-tagger.sh
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12417 (12K) [application/x-shellscript]
Saving to: ‘install-tagger.sh’

install-tagger.sh   100%[===================>]  12.13K  --.-KB/s    in 0s      

2019-06-16 11:42:20 (182 MB/s) - ‘install-tagger.sh’ saved [12417/12417]



In [5]:
!wget http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/russian.par.gz

--2019-06-16 11:42:24--  http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/russian.par.gz
Resolving www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)... 129.187.148.72, 2001:4ca0:4f01::5
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/russian.par.gz [following]
--2019-06-16 11:42:24--  https://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/data/russian.par.gz
Connecting to www.cis.uni-muenchen.de (www.cis.uni-muenchen.de)|129.187.148.72|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 115748255 (110M) [application/x-gzip]
Saving to: ‘russian.par.gz’

russian.par.gz      100%[===================>] 110.39M  10.0MB/s    in 13s     

2019-06-16 11:42:38 (8.46 MB/s) - ‘russian.par.gz’ saved [115748255/115748255]



Устанавливаем.

In [6]:
!sh install-tagger.sh


TreeTagger version for PC-Linux installed.
Tagging scripts installed.
Russian parameter file installed.
Path variables modified in tagging scripts.

You might want to add /content/cmd and /content/bin to the PATH variable so that you do not need to specify the full path to run the tagging scripts.



Предобученная модель для русского уже есть. Можно потестить.

In [7]:
!echo 'Я хочу съесть яблоко!' | cmd/tree-tagger-russian 

	reading parameters ...
	tagging ...
Я	P-1-snn	я
хочу	Vmip1s-a-e	хотеть
съесть	Vmn----a-p	съесть
яблоко	Ncnsan	яблоко
!	SENT	!
	 finished.


Но тэги тут странноватые. Чтобы оценить это все на нашем корпусе, нужно будет обучить модель заново.

In [8]:
!wget http://opencorpora.org/files/export/annot/annot.opcorpora.no_ambig_strict.xml.zip

--2019-06-16 11:51:23--  http://opencorpora.org/files/export/annot/annot.opcorpora.no_ambig_strict.xml.zip
Resolving opencorpora.org (opencorpora.org)... 148.251.2.141
Connecting to opencorpora.org (opencorpora.org)|148.251.2.141|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2194522 (2.1M) [application/zip]
Saving to: ‘annot.opcorpora.no_ambig_strict.xml.zip’

annot.opcorpora.no_ 100%[===================>]   2.09M  1.01MB/s    in 2.1s    

2019-06-16 11:51:26 (1.01 MB/s) - ‘annot.opcorpora.no_ambig_strict.xml.zip’ saved [2194522/2194522]



In [9]:
!unzip annot.opcorpora.no_ambig_strict.xml.zip

Archive:  annot.opcorpora.no_ambig_strict.xml.zip
  inflating: annot.opcorpora.no_ambig_strict.xml  


In [0]:
from lxml import etree
from collections import defaultdict

In [0]:
open_corpora = etree.fromstring(open('annot.opcorpora.no_ambig_strict.xml', 'rb').read())

Формат файлов делаем по инструкции.

In [0]:
corpus = open('corpus_train.txt', 'w')
vocab = defaultdict(set)
tags = set()

for sentence in open_corpora.xpath('//tokens'):
    length = len(sentence.xpath('token'))
    ended = False
    for i,token in enumerate(sentence.xpath('token')):
        word = token.xpath('@text')
        gram_info = token.xpath('tfr/v/l/g/@v')
        
        if (i+1)==length and gram_info[0] == 'PNCT':
            gram_info = ['SENT']
            ended = True
        
            
        corpus.write(word[0] + '\t' + ','.join(gram_info) + '\n')
        lemma = token.xpath('tfr/v/l/@t')[0]
        vocab[word[0].lower()].add((','.join(gram_info), lemma.lower()))
        tags.add(','.join(gram_info))
    
    if not ended:
        corpus.write('.\tSENT\n')

        

In [13]:
from sklearn.model_selection import KFold
f = open('corpus_train.txt', 'r')
text = f.readlines()
f.close()

filenametrain = 'corpus_train_'
filenametest = 'corpus_test_'

i = 1
kf = KFold(n_splits=4, random_state=None, shuffle=False)
for train_index, test_index in kf.split(text):
  text_train = []
  text_test = []
  for i in train_index:
    text_train.append(text[i])
  for i in test_index:
    text_test.append(text[i])
  fullfilenametrain = filenametrain + str(i) + '.txt'
  newcorpustrain = open(fullfilenametrain, 'w')
  print (fullfilenametrain)
  for el in text_train:
    newcorpustrain.write(el)
  fullfilenametest = filenametest + str(i) + '.txt'
  newcorpustest = open(fullfilenametest, 'w')
  for el in text_train:
    newcorpustest.write(el)
  i+=1
  


corpus_train_17836.txt
corpus_train_35672.txt
corpus_train_53508.txt
corpus_train_71344.txt


Словарь создаем по инструкции.

In [0]:
f = open('lexicon.txt', 'w')

for word in vocab:
    f.write(word + '\t')
    f.write('\t'.join([' '.join(pair) for pair in vocab[word]]))
    f.write('\n')
# f.write('SENT\tSENT .')
f.close()

Словарь для UNK создаем по инструкции.

In [0]:
f = open('open_class.txt', 'w')

f.write('\n'.join([tag for tag in tags if 'NOUN' in tag or 'VERB' in tag or 'ADJF' in tag]))
f.close()

**1-ый фолд**

In [16]:
 !./bin/train-tree-tagger lexicon.txt open_class.txt corpus_train_17836.txt model_oc 


train-tree-tagger -cl 2 -dtg 0.50 -sw 1.00 -ecw 0.15 -atg 1.20 lexicon.txt open_class.txt corpus_train_17836.txt model_oc

	reading the lexicon ...
		reading the tagset ...
		reading the lemmas ...
		reading the entries ...
		sorting the lexicon ...
		reading the open class tags ...
	calculating tag frequencies ...
43000	making affix tree ...
prefix lexicon: 807 nodes
suffix lexicon: 1801 nodes
	reading classes ...
	making ngram table ...
53508	26652
finished.
	making decision tree ...
90	saving parameters ...

Number of nodes: 91
Max. path length: 16

done.


In [0]:
test = open('test_corpus.txt', 'w')

test.write('\n'.join('Я хочу съесть яблоко .'.split()))

test.close()

In [24]:
!./bin/tree-tagger model_oc test_corpus.txt output.txt
!head output.txt

	reading parameters ...
	tagging ...
	 finished.
NOUN,inan,masc,sing,nomn
VERB,impf,tran,sing,1per,pres,indc
VERB,perf,tran,masc,sing,past,indc
NOUN,anim,masc,Name,sing,nomn
SENT


In [0]:
import re
testwords = open('testcorpus1.txt', 'w')
testlabels = open('testlabels1.txt', 'w')
f = open('corpus_test_17836.txt', 'r')
firsttest = f.readlines()
f.close()
for line in firsttest:
  splitline = re.split(r'\t+', line)
  testwords.write(splitline[0])
  testwords.write('\n')
  testlabels.write(splitline[1])
testwords.close()
testlabels.close()


In [30]:
!./bin/tree-tagger model_oc testcorpus1.txt testoutput1.txt

	reading parameters ...
	tagging ...
53000	 finished.


In [0]:
to = open('testoutput1.txt', 'r')
testoutput1 = to.readlines()
to.close()
tl = open('testlabels1.txt', 'r')
testlabels1 = tl.readlines()
tl.close()

In [0]:
from collections import Counter
import numpy as np

preds1 = []
mistakes1 = Counter()
tc = open('testcorpus1.txt', 'r')
testcorpus1 = tc.readlines()
tc.close()
for i in range (0,53508):
  p = int(testlabels1[i] == testoutput1[i])
  preds1.append(p)
  if not p:
    mistakes1.update([(testcorpus1[i], testlabels1[i], testoutput1[i])])

In [57]:
print(np.mean(preds1))
mistakes1.most_common(5)


0.8031582881704354


[(('Статья\n', 'NOUN,inan,femn,sing,nomn\n', 'NOUN,inan,masc,sing,nomn\n'),
  364),
 (('В\n', 'PREP\n', 'NOUN,inan,masc,sing,nomn\n'), 321),
 (('.\n', 'PNCT\n', 'SENT\n'), 299),
 (('Я\n', 'NPRO,1per,sing,nomn\n', 'NOUN,inan,masc,sing,nomn\n'), 132),
 ((',\n', 'SENT\n', 'PNCT\n'), 97)]

**2-ой фолд**

In [33]:
!./bin/train-tree-tagger lexicon.txt open_class.txt corpus_train_35672.txt model_oc1 


train-tree-tagger -cl 2 -dtg 0.50 -sw 1.00 -ecw 0.15 -atg 1.20 lexicon.txt open_class.txt corpus_train_35672.txt model_oc1

	reading the lexicon ...
		reading the tagset ...
		reading the lemmas ...
		reading the entries ...
		sorting the lexicon ...
		reading the open class tags ...
	calculating tag frequencies ...
43000	making affix tree ...
prefix lexicon: 807 nodes
suffix lexicon: 1789 nodes
	reading classes ...
	making ngram table ...
53509	26310
finished.
	making decision tree ...
96	saving parameters ...

Number of nodes: 97
Max. path length: 16

done.


In [0]:
testwords = open('testcorpus2.txt', 'w')
testlabels = open('testlabels2.txt', 'w')
f = open('corpus_test_35672.txt', 'r')
secttest = f.readlines()
f.close()
for line in secttest:
  splitline = re.split(r'\t+', line)
  testwords.write(splitline[0])
  testwords.write('\n')
  testlabels.write(splitline[1])
testwords.close()
testlabels.close()

In [53]:
!./bin/tree-tagger model_oc1 testcorpus2.txt testoutput2.txt

	reading parameters ...
	tagging ...
53000	 finished.


In [0]:
to = open('testoutput2.txt', 'r')
testoutput2 = to.readlines()
to.close()
tl = open('testlabels2.txt', 'r')
testlabels2 = tl.readlines()
tl.close()

In [0]:
preds2 = []
mistakes2 = Counter()
tc = open('testcorpus2.txt', 'r')
testcorpus2 = tc.readlines()
tc.close()
for i in range (0,53508):
  p = int(testlabels2[i] == testoutput2[i])
  preds2.append(p)
  if not p:
    mistakes2.update([(testcorpus2[i], testlabels2[i], testoutput2[i])])

In [59]:
print(np.mean(preds2))
mistakes2.most_common(5)

0.8026276444643792


[(('Статья\n', 'NOUN,inan,femn,sing,nomn\n', 'NOUN,inan,masc,sing,nomn\n'),
  365),
 (('В\n', 'PREP\n', 'NOUN,inan,masc,sing,nomn\n'), 329),
 (('.\n', 'PNCT\n', 'SENT\n'), 292),
 (('Я\n', 'NPRO,1per,sing,nomn\n', 'NOUN,inan,masc,sing,nomn\n'), 115),
 (('Не\n', 'PRCL\n', 'NOUN,inan,masc,sing,nomn\n'), 100)]

**3-ий фолд**

In [60]:
!./bin/train-tree-tagger lexicon.txt open_class.txt corpus_train_53508.txt model_oc2


train-tree-tagger -cl 2 -dtg 0.50 -sw 1.00 -ecw 0.15 -atg 1.20 lexicon.txt open_class.txt corpus_train_53508.txt model_oc2

	reading the lexicon ...
		reading the tagset ...
		reading the lemmas ...
		reading the entries ...
		sorting the lexicon ...
		reading the open class tags ...
	calculating tag frequencies ...
43000	making affix tree ...
prefix lexicon: 807 nodes
suffix lexicon: 1792 nodes
	reading classes ...
	making ngram table ...
53509	27954
finished.
	making decision tree ...
76	saving parameters ...

Number of nodes: 77
Max. path length: 15

done.


In [0]:
testwords = open('testcorpus3.txt', 'w')
testlabels = open('testlabels3.txt', 'w')
f = open('corpus_test_53508.txt', 'r')
thirdtest = f.readlines()
f.close()
for line in thirdtest:
  splitline = re.split(r'\t+', line)
  testwords.write(splitline[0])
  testwords.write('\n')
  testlabels.write(splitline[1])
testwords.close()
testlabels.close()

In [62]:
!./bin/tree-tagger model_oc2 testcorpus3.txt testoutput3.txt


	reading parameters ...
	tagging ...
53000	 finished.
PNCT
NOUN,inan,masc,sing,nomn
NOUN,inan,neut,sing,gent
PNCT
VERB,impf,tran,sing,3per,pres,indc
INFN,perf,tran
NOUN,inan,masc,sing,accs
SENT
NOUN,inan,masc,sing,nomn
PRCL


In [0]:
to = open('testoutput3.txt', 'r')
testoutput3 = to.readlines()
to.close()
tl = open('testlabels3.txt', 'r')
testlabels3 = tl.readlines()
tl.close()

In [0]:
preds3 = []
mistakes3 = Counter()
tc = open('testcorpus3.txt', 'r')
testcorpus3 = tc.readlines()
tc.close()
for i in range (0,53508):
  p = int(testlabels3[i] == testoutput3[i])
  preds3.append(p)
  if not p:
    mistakes3.update([(testcorpus3[i], testlabels3[i], testoutput3[i])])

In [66]:
print(np.mean(preds3))
mistakes3.most_common(5)

0.8039358600583091


[(('В\n', 'PREP\n', 'NOUN,inan,masc,sing,nomn\n'), 352),
 (('.\n', 'PNCT\n', 'SENT\n'), 287),
 (('Я\n', 'NPRO,1per,sing,nomn\n', 'NOUN,inan,masc,sing,nomn\n'), 143),
 (('Не\n', 'PRCL\n', 'NOUN,inan,masc,sing,nomn\n'), 101),
 (('»\n', 'SENT\n', 'PNCT\n'), 86)]

**4-ый фолд**


In [67]:
!./bin/train-tree-tagger lexicon.txt open_class.txt corpus_train_71344.txt model_oc3


train-tree-tagger -cl 2 -dtg 0.50 -sw 1.00 -ecw 0.15 -atg 1.20 lexicon.txt open_class.txt corpus_train_71344.txt model_oc3

	reading the lexicon ...
		reading the tagset ...
		reading the lemmas ...
		reading the entries ...
		sorting the lexicon ...
		reading the open class tags ...
	calculating tag frequencies ...
43000	making affix tree ...
prefix lexicon: 807 nodes
suffix lexicon: 1804 nodes
	reading classes ...
	making ngram table ...
53412	28079
finished.
	making decision tree ...
82	saving parameters ...

Number of nodes: 83
Max. path length: 14

done.


In [0]:
testwords = open('testcorpus4.txt', 'w')
testlabels = open('testlabels4.txt', 'w')
f = open('corpus_test_71344.txt', 'r')
fourtest = f.readlines()
f.close()
for line in fourtest:
  splitline = re.split(r'\t+', line)
  testwords.write(splitline[0])
  testwords.write('\n')
  testlabels.write(splitline[1])
testwords.close()
testlabels.close()

In [70]:
!./bin/tree-tagger model_oc3 testcorpus4.txt testoutput4.txt

	reading parameters ...
	tagging ...
53000	 finished.


In [0]:
to = open('testoutput4.txt', 'r')
testoutput4 = to.readlines()
to.close()
tl = open('testlabels4.txt', 'r')
testlabels4 = tl.readlines()
tl.close()

In [0]:
preds4 = []
mistakes4 = Counter()
tc = open('testcorpus4.txt', 'r')
testcorpus4 = tc.readlines()
tc.close()
for i in range (0,53412):
  p = int(testlabels4[i] == testoutput4[i])
  preds4.append(p)
  if not p:
    mistakes4.update([(testcorpus4[i], testlabels4[i], testoutput4[i])])

In [78]:
print(np.mean(preds4))
mistakes4.most_common(5)

0.8081704485883322


[(('Статья\n', 'NOUN,inan,femn,sing,nomn\n', 'NOUN,inan,masc,sing,nomn\n'),
  348),
 (('В\n', 'PREP\n', 'NOUN,inan,masc,sing,nomn\n'), 335),
 (('.\n', 'PNCT\n', 'SENT\n'), 254),
 (('Я\n', 'NPRO,1per,sing,nomn\n', 'NOUN,inan,masc,sing,nomn\n'), 134),
 ((',\n', 'SENT\n', 'PNCT\n'), 90)]

Среднее качество по всем фолдам 81%. Самые частотные ошибки это знаки препинания, которые программа не может сама рзличать. Так же слово Статья, которое выделяется как слово мужского рода. Моё предположение, что оно часто используется в сочетании со словами говорится, пишется, которые обычно идут с мужским родом.
Так же три теггер помечал слово Я как существительнео мужского рода. МНе кажется, что это произошло, из-за того что большое количество рассказчиков мужчины, от чего у программы выстроились не совсем правильные связи